src/apps/lwaftr/lwaftr.lua

```diff
@@ -379,9 +380,63 @@ local function icmpv4_incoming(lwstate, pkt)
                            lwstate.aftr_mac_b4_side, lwstate.b4_mac)
 end
 
--- TODO: correctly handle fragmented IPv4 packets
+
+local function cache_ipv4_fragment(lwstate, frag)
+   local cache = lwstate.fragment4_cache
+   local frag_id = rd16(frag.data + lwstate.l2_size + constants.o_ipv4_identification)
+   local src_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_src_addr, 8)
+   local dst_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_dst_addr, 8)
+   local src_dst = src_ip .. dst_ip
+
+   cache[src_dst] = cache[src_dst] or { n = 0 }
+   if not cache[src_dst][frag_id] then
+      cache[src_dst].n = cache[src_dst].n + 1
+      cache[src_dst][frag_id] = {}
+   end
+
+   table.insert(cache[src_dst][frag_id], frag)
+   return cache[src_dst][frag_id]
+end
+
+local function clean_ipv4_fragment_cache(lwstate, frag)
+   local cache = lwstate.fragment4_cache
+   local frag_id = rd16(frag.data + lwstate.l2_size + constants.o_ipv4_identification)
+   local src_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_src_addr, 8)
+   local dst_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_dst_addr, 8)
+   local src_dst = src_ip .. dst_ip
+
+   cache[src_dst][frag_id] = nil
+   cache[src_dst].n = cache[src_dst].n - 1
+   if cache[src_dst].n == 0 then
+      cache[src_dst] = nil
+   end
+end
+
+
 -- The incoming packet is a complete one with ethernet headers.
 local function from_inet(lwstate, pkt)
+   if fragmentv4.is_ipv4_fragment(pkt, lwstate.l2_size) then
+      local frags = cache_ipv4_fragment(lwstate, pkt)
+      local frag_status, maybe_pkt = fragmentv4.reassemble_ipv4(frags, lwstate.l2_size)
+      if frag_status == fragmentv4.REASSEMBLE_MISSING_FRAGMENT then
+         return -- Nothing useful to be done yet
+      elseif frag_status == fragmentv4.REASSEMBLE_INVALID then
+         if maybe_pkt then -- This is an ICMP packet
+            clean_ipv4_fragment_cache(lwstate, pkt)
+            if lwstate.policy_icmpv4_outgoing == lwconf.policies["DROP"] then
+               packet.free(maybe_pkt)
+            else
+               guarded_transmit(maybe_pkt, lwstate.o4)
+            end
+         end
+         return
+      else -- Reassembly was successful
+         if debug then lwdebug.print_pkt(maybe_pkt) end
+         clean_ipv4_fragment_cache(lwstate, pkt)
```

Comment:
Also, I believe it is correct to clear the cache after a successful reassembly: the returned packet from the reassembly is a new one, and the fragments have been freed, therefore there's no point in keeping references to them in the cache.

Clearing the fragment cache is not done after a successful IPv6 reassembly, so I believe we are leaking memory there. Or, if a fragment with a previously-seen identifier is received, it will be appended to a list of fragments in the cache which references (invalid?) already freed packets. Most likely behaviour would be either breakage, or in the best case fragments are deemed as invalid, and the cache cleared.